ecommerce/api.py

```diff
@@ -0,0 +1,149 @@
+            courseprice__is_valid=True,
+            course__program__programenrollment__user=user,
+        )
+    except Http404:
+        log.warning("Course run %s is not purchasable", course_key)
+        raise
+
+    # Make sure it's not already purchased
+    if Line.objects.filter(
+            order__status=Order.FULFILLED,
+            order__user=user,
+            course_key=course_run.edx_course_key,
+    ).exists():
+        log.warning("Course run %s is already purchased by user %s", course_key, user)
+        raise ValidationError("Course run {} is already purchased".format(course_key))
+
+    return course_run
+
+
+@transaction.atomic
+def create_unfulfilled_order(course_id, user):
+    """
+    Create a new Order which is not fulfilled for a purchasable course run. If course run is not purchasable,
+    it raises an Http404
+
+    Args:
+        course_id (str):
+            A course key
+        user (User):
+            The purchaser of the course run
+    """
+    course_run = get_purchasable_course_run(course_id, user)
+    price = course_run.courseprice_set.get(is_valid=True).price
+    order = Order.objects.create(
+        status=Order.CREATED,
+        total_price_paid=price,
+        user=user,
+    )
+    Line.objects.create(
+        order=order,
+        course_key=course_run.edx_course_key,
+        description='Seat for {}'.format(course_run.title),
+        price=price,
+    )
+    return order
+
+
+def generate_cybersource_sa_signature(payload):
+    """
+    Generate an HMAC SHA256 signature for the CyberSource Secure Acceptance payload
+
+    Args:
+        payload (dict): The payload to be sent to CyberSource
+    Returns:
+        str: The signature
+    """
+    # Not sure where this is documented by CyberSource but edx ecommerce uses the same algorithm
+    # https://github.com/edx/ecommerce/blob/a88d666f6362c6f9d4b6cf92b5989390e328ee8c/ecommerce/extensions/payment/processors/cybersource.py
+    keys = sorted(payload.keys())
+    message = ','.join('{}={}'.format(key, payload[key]) for key in keys)
+
+    digest = hmac.new(
+        settings.CYBERSOURCE_SECURITY_KEY.encode('utf-8'),
+        msg=message.encode('utf-8'),
+        digestmod=hashlib.sha256,
+    ).digest()
+
+    return b64encode(digest).decode('utf-8')
+
+
+def generate_cybersource_sa_payload(order):
+    """
+    Generates a payload dict to send to CyberSource for Secure Acceptance
+
+    Args:
+        order (Order): An order
+    Returns:
+        dict: the payload to send to CyberSource via Secure Acceptance
+    """
+    # http://apps.cybersource.com/library/documentation/dev_guides/Secure_Acceptance_WM/Secure_Acceptance_WM.pdf
+    # Section: API Fields
+    payload = {
+        'access_key': settings.CYBERSOURCE_ACCESS_KEY,
+        'amount': str(order.total_price_paid),
+        'consumer_id': get_social_username(order.user),
+        'currency': 'USD',
+        'locale': 'en-us',
+        # TODO
```

Comment:
I'll remove the TODO. It's to replace the two next lines with urls with valid order receipt (#904) and order cancellation pages (#905)